S21/pnklshh/refactorme.py

```diff
@@ -1,84 +1,128 @@
-      return temp
-    elif language=='ga':
-      if len(self._w)>1:
-        if (self._w[0]=='t' or self._w[0]=='n') and unicodedata.normalize('NFC', self._w)[1] in 'AEIOU\u00c1\u00c9\u00cd\u00d3\u00da':
-          temp = self._w[0]+'-'+temp[1:]
-      return temp.lower()
-    elif language=='tr':
-      temp = self._w
-      temp = temp.replace('\u0049','\u0131')
-      return temp.lower()
-    elif language=='az':
-      temp = self._w
-      temp = temp.replace('\u0049','\u0131')
-      return temp.lower()
-    elif language=='th':
-      return temp
-    elif language=='el':
-      if temp[-1]=='\u03a3':
-        self._finalSigma = True
-        temp = temp[:-1]+'\u03c2'
-      return temp.lower()
-    elif False and language=='gd':
-      # specification doesn't ask for this language to be treated differently
-      # so this will never be called
-      if len(self._w)>1:
-        if (self._w[0]=='t' or self._w[0]=='n') and self._w[1] in 'AEIOU\u00c1\u00c9\u00cd\u00d3\u00da':
-          temp = self._w[0]+'-'+temp[1:]
-      return temp.lower()
-    else:
-      return temp.lower()
+    if len(self._word)>1:
+        if (self.specialCase()):
+          self._word = self._word[0]+'-'+self._word[1:]
+    return self._word.lower()
 
   def isLenited(self):
-    language = self._l
-    if '-' in self._l:
-      i = self._l.find('-')
-      language = self._l[0:i]
+    language = self._language
+    if '-' in self._language:
+      index = self._language.find('-')
+      language = self._language[0:index]
     if language == 'ga' or language == 'gd':
-      if len(self._w) < 2:
+      if len(self._word) < 2:
         return False
       else:
-        return self._w[0].lower() in 'bcdfgmpst' and self._w[1].lower()=='h'
+        return self._word[0].lower() in 'bcdfgmpst' and self._word[1].lower()=='h'
     else:
-      raise NotImplementedError('Method only available for Irish and Scottish Gaelic')
+      raise InvalidMethodException('Method only available for Irish and Scottish Gaelic')
+
+class GreekWord(Word):
+
+  def __init__(self, word, bcpCode):
+    Word.__init__(self, word, bcpCode)
+    self._finalSigma = False
+
+  def toLower(self):
+    if self._word[-1]=='Σ':
+        self._finalSigma = True
+        self._word = self._word[:-1]+'ς'
+    return self._word.lower()
+
+class TurkishWord(Word):
+
+  def __init__(self, word, bcpCode):
+    Word.__init__(self, word, bcpCode)
+
+  def toLower(self):
+    self._word = self._word.replace('I','ı')
+    return self._word.lower()
+
+class AzerbaijaniWord(Word):
+
+  def __init__(self, word, bcpCode):
+    Word.__init__(self, word, bcpCode)
+
+  def toLower(self):
+    self._word = self._word.replace('I','ı')
+    return self._word.lower()
+
+class InvalidLanguageException(Exception):
+    def __init__(self, message):
+        super().__init__(message)
+
+class InvalidMethodException(Exception):
+    def __init__(self, message):            
+        super().__init__(message)
+
+class Language:
```

Comment:
Nice idea.  Might like "BCP47" in the name of this class.  Also, by including the skipLanguageList your making this class specific to the lowercasing problem, and in reality it could/should be more generic.